In [1]:
import numpy as np
import pandas as pd
import pathlib as pl
from tqdm import tqdm
import matplotlib.pyplot as plt



selector = 5

In [2]:
current_path        = pl.Path.cwd()
datapath            = current_path.joinpath('Modelos')

if selector == 0:
    folder_path = datapath.joinpath('1. U-Net')
elif selector == 1:
    folder_path = datapath.joinpath('2. SegNet')
elif selector == 2:
    folder_path = datapath.joinpath('3. U-Net_VGG16')
elif selector == 3:
    folder_path = datapath.joinpath('4. U-Net_ResNet50')
elif selector == 4:
    folder_path = datapath.joinpath('5. FNP_Resnet50')
elif selector == 5:
    folder_path = datapath.joinpath('6. LinkNet_VGG16')

cvs_file    = list(folder_path.glob('*.csv'))
df          = pd.read_csv(cvs_file[0])
print(current_path)
print(datapath)
print(folder_path)

d:\Trabajo de Maestria\segmentacion_maestria
d:\Trabajo de Maestria\segmentacion_maestria\Modelos
d:\Trabajo de Maestria\segmentacion_maestria\Modelos\6. LinkNet_VGG16


In [3]:
df.head()

accuracy  dice_score  f1-score  f2-score  iou_score      loss  precision  \
0  0.887471    0.424559  0.424559  0.425003   0.349180  0.391605   0.435563   
1  0.941959    0.532113  0.532113  0.535314   0.460354  0.168530   0.528806   
2  0.945392    0.557667  0.557667  0.560843   0.483011  0.148097   0.554692   
3  0.942889    0.554910  0.554910  0.557852   0.479264  0.152720   0.553922   
4  0.946683    0.570669  0.570669  0.573272   0.493879  0.140579   0.568611   

     recall  val_accuracy  val_dice_score  val_f1-score  val_f2-score  \
0  0.432377      0.879810        0.412829      0.412829      0.414905   
1  0.538170      0.854627        0.377917      0.377917      0.379048   
2  0.563715      0.845239        0.385992      0.385992      0.385517   
3  0.560982      0.856354        0.362518      0.362518      0.369383   
4  0.575710      0.860894        0.386921      0.386921      0.389690   

   val_iou_score  val_loss  val_precision  val_recall  
0       0.337983  0.926698       0.425727    0.439930  
1       0.325183  1.111715       0.407250    0.395855  
2       0.328097  2.008249       0.432308    0.402366  
3       0.323930  1.621668       0.463998    0.387044  
4       0.340476  1.381457       0.498604    0.406197

In [4]:

# df = df.round(4) # esto se puede hacer al final pero dejarlo por el momento
df.head()


accuracy  dice_score  f1-score  f2-score  iou_score      loss  precision  \
0  0.887471    0.424559  0.424559  0.425003   0.349180  0.391605   0.435563   
1  0.941959    0.532113  0.532113  0.535314   0.460354  0.168530   0.528806   
2  0.945392    0.557667  0.557667  0.560843   0.483011  0.148097   0.554692   
3  0.942889    0.554910  0.554910  0.557852   0.479264  0.152720   0.553922   
4  0.946683    0.570669  0.570669  0.573272   0.493879  0.140579   0.568611   

     recall  val_accuracy  val_dice_score  val_f1-score  val_f2-score  \
0  0.432377      0.879810        0.412829      0.412829      0.414905   
1  0.538170      0.854627        0.377917      0.377917      0.379048   
2  0.563715      0.845239        0.385992      0.385992      0.385517   
3  0.560982      0.856354        0.362518      0.362518      0.369383   
4  0.575710      0.860894        0.386921      0.386921      0.389690   

   val_iou_score  val_loss  val_precision  val_recall  
0       0.337983  0.926698       0.425727    0.439930  
1       0.325183  1.111715       0.407250    0.395855  
2       0.328097  2.008249       0.432308    0.402366  
3       0.323930  1.621668       0.463998    0.387044  
4       0.340476  1.381457       0.498604    0.406197

In [5]:
# Crear una copia del slice del DataFrame original para evitar el warning
df_train = df.iloc[:, :8].copy()

df_train['puntaje'] = (
    df_train['accuracy'] +
    df_train['dice_score'] +
    df_train['f1-score'] +
    df_train['f2-score'] +
    df_train['iou_score'] +
    df_train['precision'] +
    df_train['recall'] +
    (1 - df_train['loss'])
    )

# Mostrar las primeras filas del DataFrame modificado
df_train.head()

accuracy  dice_score  f1-score  f2-score  iou_score      loss  precision  \
0  0.887471    0.424559  0.424559  0.425003   0.349180  0.391605   0.435563   
1  0.941959    0.532113  0.532113  0.535314   0.460354  0.168530   0.528806   
2  0.945392    0.557667  0.557667  0.560843   0.483011  0.148097   0.554692   
3  0.942889    0.554910  0.554910  0.557852   0.479264  0.152720   0.553922   
4  0.946683    0.570669  0.570669  0.573272   0.493879  0.140579   0.568611   

     recall   puntaje  
0  0.432377  3.987107  
1  0.538170  4.900299  
2  0.563715  5.074890  
3  0.560982  5.052010  
4  0.575710  5.158914

In [6]:
df_train = df_train.round(4)
df_train.iloc[[-1]]

accuracy  dice_score  f1-score  f2-score  iou_score    loss  precision  \
249     0.943      0.5613    0.5613    0.5656     0.4866  0.1653     0.5674   

     recall  puntaje  
249  0.5729   5.0929

0.943 & 0.5613 & 0.5613 & 0.5656 & 0.4866 & 0.1653 & 0.5674 & 0.5729 & 5.0929

In [7]:
df_val = df.iloc[:, 8:16].copy()

df_val['puntaje'] = (
    df_val['val_accuracy'] +
    df_val['val_dice_score'] +
    df_val['val_f1-score'] +
    df_val['val_f2-score'] +
    df_val['val_iou_score'] +
    df_val['val_precision'] +
    df_val['val_recall'] +
    (1 - df_val['val_loss'])
    )

df_val.head()

val_accuracy  val_dice_score  val_f1-score  val_f2-score  val_iou_score  \
0      0.879810        0.412829      0.412829      0.414905       0.337983   
1      0.854627        0.377917      0.377917      0.379048       0.325183   
2      0.845239        0.385992      0.385992      0.385517       0.328097   
3      0.856354        0.362518      0.362518      0.369383       0.323930   
4      0.860894        0.386921      0.386921      0.389690       0.340476   

   val_loss  val_precision  val_recall   puntaje  
0  0.926698       0.425727    0.439930  3.397317  
1  1.111715       0.407250    0.395855  3.006082  
2  2.008249       0.432308    0.402366  2.157261  
3  1.621668       0.463998    0.387044  2.504077  
4  1.381457       0.498604    0.406197  2.888246

In [8]:
df_val = df_val.round(4)
df_val.iloc[[-1]]

val_accuracy  val_dice_score  val_f1-score  val_f2-score  val_iou_score  \
249         0.935          0.4935        0.4935        0.4879         0.4391   

     val_loss  val_precision  val_recall  puntaje  
249    0.2311          0.562      0.4981    4.678

0.935 & 0.4935 & 0.4935 & 0.4879 & 0.4391 & 0.2311 & 0.562 & 0.4981 & 4.678

La arquitectura \textit{LinkNet con VGG16} ha sido evaluada en este estudio para determinar su desempeño en la segmentación semántica de imágenes. Las métricas clave de rendimiento, han sido calculadas tanto para los datos de entrenamiento como de prueba, y se presentan en las Tablas \ref{tab:tabla-linknet-vgg16-1} y \ref{tab:tabla-linknet-vgg16-2}.

\begin{table}[h]
\centering
\begin{tabular}{|l|l|l|l|l|l|l|l|l|}
\hline
\textit{accuracy} & \textit{dice score} & \textit{f1-score} & \textit{f2-score} & \textit{IOU score} & \textit{loss} & \textit{precision} & \textit{recall} & \textbf{puntaje}\\
\hline
0.943 & 0.5613 & 0.5613 & 0.5656 & 0.4866 & 0.1653 & 0.5674 & 0.5729 & 5.0929\\
\hline
\end{tabular}
\caption{Tabla de resultados de LinkNet con VGG16 en Train}
\label{tab:tabla-linknet-vgg16-1}
\end{table}

La Tabla \ref{tab:tabla-linknet-vgg16-2} muestra que el modelo alcanzó un puntaje de 5.0929, un \textit{accuracy} de 0.943, indicando una alta tasa de acierto en la clasificación de píxeles. El \textit{dice\_coefficient} de 0.5613 y el \textit{iou} de 0.4866 con una pobre superposición entre las predicciones del modelo y las etiquetas verdaderas, demostrando la incapacidad del modelo para identificar con precisión las regiones de interés. Los valores de \textit{precision} y \textit{recall}, 0.5674 y 0.5729 respectivamente, destacan la ineficacia del modelo en la detección de píxeles relevantes. El \textit{loss} registrado fue de 0.5674, lo cual es relativamente alto, indicando un mal desempeño general del modelo durante el entrenamiento. Los resultados obtenidos en la evaluación de LinkNet con VGG16 en entrenamiento se encuentran por debajo de lo resultados obtenidos con \textit{U-Net}

\begin{table}[h]
\centering
\begin{tabular}{|l|l|l|l|l|l|l|l|l|}
\hline
\textit{accuracy} & \textit{dice score} & \textit{f1-score} & \textit{f2-score} & \textit{IOU score} & \textit{loss} & \textit{precision} & \textit{recall} & \textbf{puntaje}\\
\hline
0.935 & 0.4935 & 0.4935 & 0.4879 & 0.4391 & 0.2311 & 0.562 & 0.4981 & 4.678\\
\hline
\end{tabular}
\caption{Tabla de resultados de LinkNet con VGG16 en Test}
\label{tab:tabla-linknet-vgg16-2}
\end{table}

Los resultados obtenidos en la evaluación de LinkNet con VGG16 en test, muestran que el modelo alcanzó un puntaje de 4.678, un \textit{accuracy} de 0.935, indicando una alta tasa de acierto en la clasificación de píxeles. El \textit{dice\_coefficient} de 0.4935 y el \textit{iou} de 0.4391 con una pobre superposición entre las predicciones del modelo y las etiquetas verdaderas, demostrando la incapacidad del modelo para identificar con precisión las regiones de interés. Los valores de \textit{precision} y \textit{recall}, 0.562 y 0.4981 respectivamente, destacan la ineficacia del modelo en la detección de píxeles relevantes. El \textit{loss} registrado fue de 0.2311, lo cual es relativamente alto, indicando un mal desempeño general del modelo durante el entrenamiento. Los resultados obtenidos en la evaluación de LinkNet con VGG16 en test se encuentran por debajo de lo resultados obtenidos con \textit{U-Net}